### ISSUE TITLE
come up w/ desired information for new status pages

### ISSUE THREAD

**Issue by jamesturk** [openstates/admintools#1]
to decide scope of status page work

- [ ] decide what we want to see prominently & what would be useful
- [ ] which things should be done in a 'reports' phase and which should be calculated on the fly?

**Comment by hiteshgarg14** [comment]
Please give some feedback about any edition/deletion/updating etc. 

Status Page

1.  Last scrape Run date (YYYY-MM-DD) with color code:- (1 Column)
       -  Green - Success
       -  Red - Fail
       -  ??

2.  Organization:- (3 Columns instead of 5)
       -  Instead of  `upper`, `lower`, & `joint` count we can have a single column `count`. When a user hovers over `count`, show a popup containing more information about `upper`, `lower`, & `joint`. And if these 3 counts need attention then there will be red color on that particular column else green color.
       -  Numbers of `empty` orgs
       -  % Members with leg_id
       -  ??

3.  People:- (5/6 Columns instead of 6)
       - `Vacant` and `Over_Filled` seats count can me merged into one ?? (like we have `mouseover ` technique)
      - `Email`, `Phone Number`, `URL` should have separate columns
      - `Photo` Column can be replaced by `Address`.
      - ??

4.  Bills:- ( 7 Columns instead of 8)
      - `Total` number of bills. When a user hovers over this column then it will show the count of `upper`, `lower` etc. 
      - `No Actions`, `No Sponsors`, `No Versions`, `With Subjects`, `Dup. Versions`, `Sponsor ids`
`No Votes`
     - ??

5.  Votes:- ( 3 Columns instead of 2)
     - Since in Pupa Conversion we still have `other` votes counts instead of `Absent or Voting`/ `Non-Voting` in many scrapers. So we should continue this we can have `mouseover` technique to see other types of voting.
      - Votes % of associated bills with them
      - We can continue leg_ids field
      - ??

**Comment by jamesturk** [comment]
bear with me here, as this is quite the different approach from what we have now, but here's an idea I've been kicking around.

right now the /admin/ page is pretty cluttered as you look at it and see maybe 60% green, 20% red and 20% yellow, someone looking at the page wouldn't know where to begin in addressing issues, so I'm not sure it does much for us

warning: NH district 7 does not have any members
error: NC district 2 has 2 members, should have 1
warning: 23 legislators in OH do not have photos
error: KS does not have any committees
warning: 6 bills in NE have no sponsors

the admin page could then be much simpler:


state | last run | legislator status | committee status | bill status | vote status
----- | -------- | ----------------| ------------------| -----------| ----------
CA    | ✅ 2017-05-30 | ✅ | ✅ | ✅ | ✅
NC    | ✅ 2017-05-30 | 0 ⚠️ 5 ❌  | ✅ | ✅ |200 ⚠️ 0 ❌
WY    | ❌ 2017-05-23 | 0 ⚠️ 5 ❌  | 12 ⚠️ 5 ❌ | ✅ |200 ⚠️ 0 ❌

**Comment by jamesturk** [comment]
I'll also add this, there's some interesting data available in pupa's saved runs now, we save how many objects were updated/inserted/not modified in each run.  It could be really interesting to surface somehow, especially if anomalies were easy to spot

**Comment by hiteshgarg14** [comment]
Instead of defining warning type codes. 
1. If the user wants to apply filters in all states then we can put options in `navbar`. like this:- 

![screenshot from 2017-05-31 22 04 13](https://cloud.githubusercontent.com/assets/19409253/26642968/6080062e-464d-11e7-86de-a8b3de5a1e9f.png)

when a user selects an option then only that column will be displayed in the table with appropriate color coding. 

2. If the user wants to apply filters in a particular state or in multiple states then he can select states and we can have the same type of `dropdown` on each row attached to state name
![screenshot from 2017-05-31 22 18 43](https://cloud.githubusercontent.com/assets/19409253/26643494/218f39b0-464f-11e7-8879-fccee6404362.png)
![screenshot from 2017-05-31 22 24 52](https://cloud.githubusercontent.com/assets/19409253/26643747/02a3e3e2-4650-11e7-9241-c0d034da7f93.png)

**Comment by jamesturk** [comment]
I don't think per-state filters are going to be needed, I think for our requirements we can say that a person is either interested in looking at an overview of states (/admin/) or a specific state (/admin/nc/) - the complexity of the UI if we go into multi-state filters is (as you noted) going to get hard.  (and I think, rarely used)

What I'm trying to rethink with my approach is the usefulness of a grid of 1000+ boxes at once.  Even if I can opt to hide some, the /admin/ view just isn't overly useful in my experience.  (having written it myself I have nobody to blame)

Taking a step back, what do we want to know?  Maybe we can make a list and then figure out what the right presentation is.

People
- are any legislative seats empty?
- are any legislative seats overfilled?
- are any legislators missing key contact information that we can collect (phone, email, etc.)
- any any legislators missing photos?

Committees
- are there empty committees?
- are committees missing?
- are there committees that have members that haven't been name-matched?

Bills/Votes
- are there bills that are missing (generally required) information (sponsors, actions, versions)
- are there bills with sponsors that haven't been name-matched?
- are there votes with voters that haven't been name-matched?
- are there votes with bad counts?

So that'd at least kill things from the table like counts, subjects, etc.

My reasoning from here to think about storing entries for each as a DataQualityIssue or something instead of just a bunch of counts/etc. is that these DataQualityIssues can then be created in other ways.
- via an automated report process as described here
- the user-feedback portion could create a DataQualityIssue when the user fills out a form
- we could even hypothetically eventually expose an API to submit new data quality issues so that third party apps can push issues upstream

Upsides:
 - extensible, could come up w/ generic interface for data quality issues
 - looking at different aggregations of the issues (by state, by type, by specific object) would be possible via cheap DB lookup
 - possible to treat future sources of DQ errors in same manner as those automatically noted by this process

Downsides:
 - as with any approach that caches this information in a DB table instead of generating it on the fly, possible for system to be out of sync (e.g. it says 40 duplicate legislators but we've merged them all)


Definitely open to hearing more thoughts on the downsides here.  The issue of caching all of this info is a real one, and frustrating right now w/ billy (where a successful run has to happen before the numbers in /admin/ are updated again for a state).  That said, unless we drastically simplify what we show on /admin/ altogether I'm not sure we have a way to do this without the need for *some* sort of caching- each cell is at least one db query, so we'd be talking about 1000+ queries per render of /admin/

**Comment by hiteshgarg14** [comment]
Yeah! I totally agree with your concern that "what is the usefulness of a grid of 1000+ boxes at once and the problem with the caching approach". So, Why don't we just focus on calculating the things on fly for the status page. In this way, we will have a page which is fully updated with the current database.

What about 1000+ queries to the database every time user hits /admin/ view?
My answer to this question is that since Django queries are lazy. so we can use that to not actually hit the whole database and query each and every entry. we can use ` QuerySet.count()` / `QuerySet.exists()`  type of things which are efficient enough. 

we can reframe the status page according to three colors `Red`, `Yellow` & `Green`. for example, If there are empty committees. then we can use `.exists()` to know if this is `True` or `False` or we can use `.count()` either. 

I know that `.count()` is too not that efficient. but we can take help form [here](https://wiki.postgresql.org/wiki/Count_estimate)

So my basic point is to make a status page just for the True/False kind of info related to various fields with some count(for displaying warnings)

just an idea wanted to share :)

**Comment by hiteshgarg14** [comment]
1. Missing Email ID
     - name = `email`
     - description = person with missing `email id`
     - warning
     - `Person` OCD ID     (Jurisdiction OCD ID ??)
2. Missing Address
     - name = `address`
     - description = person with missing `address`
     - warning
     - `Person` OCD ID 
3. Missing Phone
     - name = `phone`
     - description = person with missing `phone`
     - warning
     - `Person` OCD ID 
4. Missing Photo URL
     - name = `image`
     - description = person with missing `image`
     - warning
     - `Person` OCD ID 
5. Org without any member
     - name = `no_memberships`
     - description = A organization with no memberships. 
     - error
     - `Organization` OCD ID    
6. Org members with Person ID = None
     - name = `unmatched_person`
     - description = an org has a person but not related to any person in database.
     - warning
     - `Membership` OCD ID 
7. Vote Events
     - name = `voteevent_missing_bill`
     - description = vote events with no bills associated with them
     - warning
     - `VoteEvent` OCD ID 
8. Vote Event
    - name = `unmatched_voter`
    - description = Any vote event which has a voter with Person ID == None 
    - warning
    - `VoteEvent` & `Person` ODC IDs
9. Vote Event
    - name = `missing_voters`
    - description = Don’t have any `PERSON VOTES`
    - error
    - `VoteEvent` ODC IDs 
10. Vote Event
    - name = `missing_votes`
    - description = Don’t have any type of vote count 
    - error
    - `VoteEvent` ODC ID
11. Vote Event
    - name = `bad_votes`
    - description = With Bad count
    - warning
    - `VoteEvent` ODC ID    
12. Bills
    - name = `no_actions`
    - description = With no actions
    - error
    - `Bill` ODC ID 
13. Bills
    - name = `no_sponsors`
    - description = With no sponsors
    - warning
    - `Bill` ODC ID 
14. Bills
    - name = `unmatched_person_sponsor`
    - description = Have a sponsor with Person ID == None
    - warning
    - `Bill` & `Person` ODC IDs 
15. Bills
    - name = `no_versions`
    - description = With no versions
    - error
    - `Bill`  ODC ID
16. Bills
    - name = `unmatched_org_sponsor`
    - description = Have a sponsor with Org ID == None
    - warning
    - `Bill` & `Organization` ODC IDs 
17. Jurisdiction
    - name = `empty_seats`
    - description = has empty seats
    - warning
    - `Jurisdiction` ODC ID
18. Jurisdiction
    - name = `overfilled_seats`
    - description = has overfilled seats
    - error
    - `Jurisdiction` ODC ID

**Comment by jamesturk** [comment]
going to close this issue for now as I think we covered it 👍
